### PR TITLE
New flag 'ingest-from'

### PR DESCRIPTION
**Pull request by fkaleo**
New flag 'ingest-from' that instructs metrictank to discard any point of a specific org id that does not belong to chunks starting after a given timestamp.
e.g. 20:67031 with a chunk span of 10 will discard all points with org id 20 and timestamps lesser than 67040

### CODE REVIEW

**Comment by woodsaj** [comment]
why limit this to only 1 orgId?  Whould it not be better to just have AggMetrics use a map of orgId's to ingestAfter timstamp. eg  `ingestAfter map[int32]int32`

Users can then pass a list of org:ts pairs via the config.  eg
`metrictank --ingest-after=1:1562339919,1234:1562339919,45:1562339919`

**Comment by fkaleo** [comment]
> why limit this to only 1 orgId? Whould it not be better to just have AggMetrics use a map of orgId's to ingestAfter timstamp. eg `ingestAfter map[int32]int32`
> 
> Users can then pass a list of org:ts pairs via the config. eg
> `metrictank --ingest-after=1:1562339919,1234:1562339919,45:1562339919`

The only reason was the fact that we did not need it and kept the code a little simpler. We can always add it later.

**Comment by fkaleo** [comment]
> > why limit this to only 1 orgId? Whould it not be better to just have AggMetrics use a map of orgId's to ingestAfter timstamp. eg `ingestAfter map[int32]int32`
> > Users can then pass a list of org:ts pairs via the config. eg
> > `metrictank --ingest-after=1:1562339919,1234:1562339919,45:1562339919`
> 
> The only reason was the fact that we did not need it and kept the code a little simpler. We can always add it later.

Using a map is probably less code though; let me try that.

File: mdata/aggmetric.go [hunk 1]
```diff
@@ -422,6 +430,12 @@ func (a *AggMetric) persist(pos int) {
 
 // don't ever call with a ts of 0, cause we use 0 to mean not initialized!
 func (a *AggMetric) Add(ts uint32, val float64) {
+	if ts < a.ingestAfterNextChunkT0 {
+		// TODO: add metric to keep track of the # of points discarded
+		log.Debugf("AM: discarding metric <%d,%f>: does not belong to a chunk starting after ingest-after. First chunk considered starts at %d", ts, val, a.ingestAfterNextChunkT0)
```
[the remark this answers is not in the document]
Comment (fkaleo) [reply mdata/aggmetric.go]: I'm curious now: `if logger.IsLevelEnabled(WarnLevel)` is called by `log.Debugf` albeit one function call deeper and with more argument passing: did we notice significant perf improvements with it?
In any case, I added the `if` in the latest commit.

**Comment by fkaleo** [comment]
> I think this might need to be rebased on master

Done

File: mdata/aggmetric.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (shanson7) [reply mdata/aggmetric.go]: Yes. You can see my comment in https://github.com/grafana/metrictank/pull/1269

tl;dr The go compiler escape analysis reveals that these lines cause the arguments to escape to the heap, so they result in allocations when called.

[the remark this answers is not in the document]
Comment (fkaleo) [reply mdata/aggmetric.go]: awesome Sean!

File: cmd/metrictank/metrictank.go [hunk 2]
```diff
@@ -59,6 +60,7 @@ var (
 
 	// Data:
 	dropFirstChunk    = flag.Bool("drop-first-chunk", false, "forego persisting of first received (and typically incomplete) chunk")
+	ingestAfterStr    = flag.String("ingest-after", "", "discards data from an org id received before the next chunk boundary after the specified unix timestamp (in seconds since 1 Jan 1970); syntax: org-id:timestamp")
```
Comment (Dieterbe) [inline cmd/metrictank/metrictank.go]: 1) this is not true in case the specified timestamp is a chunk boundary (than it is not the next chunk)
2) the description of what a unix timestamp is incomplete (UTC timezone), but also this is not the place to explain a unix timestamp
3) would it make sense to have multiple rules for multiple orgs? I thought it did
4) the message is clunky. can we simplify it?

Comment (fkaleo) [reply cmd/metrictank/metrictank.go]: >     1. this is not true in case the specified timestamp is a chunk boundary (than it is not the next chunk)

I thought it was the case given how `ingestAfterNextChunkT0` is computed in `aggmetric.go`:

`ingestAfterNextChunkT0 = uint32(ingestAfter-(ingestAfter%int64(ret.ChunkSpan))) + ret.ChunkSpan`

> 
>     2. the description of what a unix timestamp is incomplete (UTC timezone), but also this is not the place to explain a unix timestamp

Removed.

> 
>     3. would it make sense to have multiple rules for multiple orgs (eg as csv) ? I thought it did

Implemented (546e9aa).

> 
>     4. the message is clunky. can we simplify it?

I took a stab at simplifying it, let me know what you think (689b147).

Comment (Dieterbe) [reply cmd/metrictank/metrictank.go]: > I thought it was the case given how ingestAfterNextChunkT0 is computed in aggmetric.go:

you are correct!

However, I imagine the typical use for a flag like this will be to pass a chunk boundary timestamp, so something like:
```
~ ❯❯❯ TZ=UTC date -d '06:00'
Wed 07 Aug 2019 06:00:00 AM UTC
~ ❯❯❯ TZ=UTC date -d '06:00' +%s
1565157600
~ ❯❯❯ TZ=UTC date -d 'tomorrow 00:00'
Thu 08 Aug 2019 12:00:00 AM UTC
~ ❯❯❯ TZ=UTC date -d 'tomorrow 00:00' +%s
1565222400
```

I don't imagine people would pass a value such as 
```
~ ❯❯❯ TZ=UTC date -d '23:59:59' +%s
1565222399
```

Thus, assuming this hunch is correct and people would pass a clean chunk boundary, i would argue it's confusing if ingestion is only for data that is 1 chunk "after" the timestamp i specified.
Thus perhaps instead of the formula you use, we should use `mdata.AggBoundary()` and change the name "ingest-after" to "ingest-from", describing that is is to ingest data "from this timestamp and onwards" rathar than strictly "after".

WDYT?

Comment (Dieterbe) [reply cmd/metrictank/metrictank.go]: > I took a stab at simplifying it, let me know what you think (689b147).

 :+1:

File: mdata/aggmetric.go [hunk 3]
```diff
@@ -31,39 +31,47 @@ type AggMetric struct {
...
-	firstTs         uint32 // timestamp of first point seen
+	key                    schema.AMKey
+	rob                    *ReorderBuffer
+	currentChunkPos        int    // Chunks[CurrentChunkPos] is active. Others are finished. Only valid when len(chunks) > 0, e.g. when data has been written (excl ROB data)
+	numChunks              uint32 // max size of the circular buffer
+	chunkSpan              uint32 // span of individual chunks in seconds
+	chunks                 []*chunk.Chunk
+	aggregators            []*Aggregator
+	dropFirstChunk         bool
+	ingestAfterNextChunkT0 uint32
```
Comment (Dieterbe) [inline mdata/aggmetric.go]: another benefit of my suggestion (see other comment) is if the user input is `ingestFrom`, then this var can be `ingestFromT0` and all the alignment here stays the same ;)

File: util/flag_parsers_test.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+		{"missing timestamp", "0", 0, 0, true},
+		{"text no number 1", "text", 0, 0, true},
+		{"text no number 2", "1:text", 0, 0, true},
+		{"orgid overflow", "10000000000:0", 0, 0, true},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			defer func() {
+				r := recover()
```
Comment (Dieterbe) [inline util/flag_parsers_test.go]: have a look at : https://golang.org/src/regexp/regexp.go?s=9796:9832#L259
it seems a go best practice to have Must* style functions call regular functions that return an error.
that way in unit tests you can call the non-Must (non-panicing) version and just check for errors, rather than needing recover logic. this seems cleaner and simpler.

Comment (fkaleo) [reply util/flag_parsers_test.go]: > have a look at : https://golang.org/src/regexp/regexp.go?s=9796:9832#L259
> it seems a go best practice to have Must* style functions call regular functions that return an error.
> that way in unit tests you can call the non-Must (non-panicing) version and just check for errors, rather than needing recover logic. this seems cleaner and simpler.

Done in a1c8bf1

File: mdata/aggmetric.go (hunk 3 above)
Comment (fkaleo) [reply mdata/aggmetric.go]: > another benefit of my suggestion (see other comment) is if the user input is `ingestFrom`, then this var can be `ingestFromT0` and all the alignment here stays the same ;)

Implemented with 3bb1d5f

File: cmd/metrictank/metrictank.go (hunk 2 above)
Comment (fkaleo) [reply cmd/metrictank/metrictank.go]: > > I thought it was the case given how ingestAfterNextChunkT0 is computed in aggmetric.go:
> 
> you are correct!
> 
> However, I imagine the typical use for a flag like this will be to pass a chunk boundary timestamp (or, a common enough use), so something like:
> 
> ```
> ~ ❯❯❯ TZ=UTC date -d '06:00'
> Wed 07 Aug 2019 06:00:00 AM UTC
> ~ ❯❯❯ TZ=UTC date -d '06:00' +%s
> 1565157600
> ~ ❯❯❯ TZ=UTC date -d 'tomorrow 00:00'
> Thu 08 Aug 2019 12:00:00 AM UTC
> ~ ❯❯❯ TZ=UTC date -d 'tomorrow 00:00' +%s
> 1565222400
> ```
> 
> I don't imagine people would often pass a value such as
> 
> ```
> ~ ❯❯❯ TZ=UTC date -d '23:59:59' +%s
> 1565222399
> ```
> 
> Thus, assuming this hunch is correct and people would commonly pass a clean chunk boundary, i would argue it's confusing if ingestion is only for data that is 1 chunk "after" the timestamp i specified.
> Thus perhaps instead of the formula you use, we should use [`mdata.AggBoundary()`](https://godoc.org/github.com/grafana/metrictank/mdata#AggBoundary) and change the name "ingest-after" to "ingest-from", describing that is is to ingest data "from this timestamp and onwards" rathar than strictly "after".
> 
> WDYT?


Implemented with commits 67e3d6a and 3a498f4.

File: mdata/aggmetric.go [hunk 5]
```diff
@@ -58,8 +58,8 @@ func NewAggMetric(store Store, cachePusher cache.CachePusher, key schema.AMKey,
 
 	ingestFromT0 := uint32(0)
 	if ingestFrom > 0 {
-		// compute start of next chunk after ingestFrom
-		ingestFromT0 = uint32(ingestFrom-(ingestFrom%int64(ret.ChunkSpan))) + ret.ChunkSpan
+		// compute next chunk boundary starting from 'ingestFrom'
```
Comment (Dieterbe) [inline mdata/aggmetric.go]: the word "next" seems confusing here.

Comment (fkaleo) [reply mdata/aggmetric.go]: > the word "next" seems confusing here.

Yeah, I don't know how to phrase it much better :/ And `AggBoundary` uses the word next too :/
`// AggBoundary returns ts if it is a boundary, or the next boundary otherwise.`

Should I remove the comment altogether?

Comment (Dieterbe) [reply mdata/aggmetric.go]: what i meant is that "next" does not apply in all cases.  AggBoundary comment has the appropriate nuance. anyway i'll give it a shot, no worries.

Comment (fkaleo) [reply mdata/aggmetric.go]: > what i meant is that "next" does not apply in all cases. AggBoundary comment has the appropriate nuance. anyway i'll give it a shot, no worries.

you're right

File: mdata/aggregator_test.go [hunk 6]
```diff
@@ -91,7 +91,35 @@ func TestAggregator(t *testing.T) {
...
+
+	// chunkspan is 120, ingestFrom = 120 means points before chunk starting at 120 are discarded
+	agg = NewAggregator(mockstore, &cache.MockCache{}, test.GetAMKey(1), ret, aggs, false, 120)
+	agg.Add(100, 123.4)
+	agg.Add(110, 5)
+	agg.Add(130, 130)
+	expected = []schema.Point{
+		{Val: 5, Ts: 120},
+	}
+	compare("simple-min-ingest-from-on-chunk-boundary", agg.minMetric, expected)
```
Comment (Dieterbe) [inline mdata/aggregator_test.go]: too tired now, but tomorrow i'ld like to look into this one last thing, about the timestamps going 130->120 and how that may or should not affect things.

**Comment by Dieterbe** [comment]
LGTM. just need to look at one last thing tomorrow.

File: mdata/aggregator_test.go (hunk 6 above)
Comment (fkaleo) [reply mdata/aggregator_test.go]: > LGTM. just need to look at one last thing tomorrow.

Thank you! Your changes look great.

File: mdata/aggmetric_test.go [hunk 7]
```diff
@@ -302,6 +302,34 @@ func TestAggMetricDropFirstChunk(t *testing.T) {
...
+	m.Add(22, 22)
+	m.Add(30, 30)
+	m.Add(31, 31)
+	m.Add(32, 32)
+	m.Add(40, 40)
+	itgens, err := mockstore.Search(test.NewContext(), test.GetAMKey(42), 0, 0, 1000)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(itgens) != 1 || itgens[0].T0 != 30 {
```
Comment (Dieterbe) [inline mdata/aggmetric_test.go]: what about point with ts=40 ? wouldn't that be in a 2nd chunk?

File: mdata/aggregator_test.go [hunk 8]
```diff
@@ -91,7 +91,35 @@ func TestAggregator(t *testing.T) {
...
+	expected = []schema.Point{}
+	compare("simple-min-ingest-from-all-before-next-chunk", agg.minMetric, expected)
+
+	// chunkspan is 120, ingestFrom = 115 means points before chunk starting at 120 are discarded
+	agg = NewAggregator(mockstore, &cache.MockCache{}, test.GetAMKey(1), ret, aggs, false, 115)
+	agg.Add(100, 123.4)
+	agg.Add(110, 5)
+	agg.Add(130, 130)
+	expected = []schema.Point{
+		{Val: 5, Ts: 120},
```
Comment (Dieterbe) [inline mdata/aggregator_test.go]: where did point ts=130 go?

Comment (fkaleo) [reply mdata/aggregator_test.go]: Comments in 3b4625e should answer that.

File: mdata/aggmetric_test.go (hunk 7 above)
Comment (fkaleo) [reply mdata/aggmetric_test.go]: Comments in 3b4625e should answer that.

**Comment by fkaleo** [comment]
> LGTM
> Please have a close look at the last 2 commits I added.
> If they make sense, merge this.
> If not, let's talk.

They make sense. I'm adding test cases to cover them.

**Comment by Dieterbe** [comment]
I already covered that in the test case i added.
if we need any more tests, it should be one that uses an actual AggMetric with aggregators on it, and confirms the right points make it in to the aggregator (points in the last aggspan before the first t0)

**Comment by fkaleo** [comment]
> I already covered that in the test case i added.
> if we need any more tests, it should be one that uses an actual AggMetric with aggregators on it, and confirms the right points make it in to the aggregator (points in the last aggspan before the first t0)

The test case added does not fail when removing the code changes.
And yep, I'm adding a test to an AggMetric and discovering that in fact more code changes need to happen to make it pass, I think.

**Comment by fkaleo** [comment]
Please take a look at the 2 additional tests and the change in code and let me know what you think.

File: mdata/aggmetric_test.go [hunk 9]
```diff
@@ -302,6 +308,156 @@ func TestAggMetricDropFirstChunk(t *testing.T) {
...
+		t.Fatalf("expected 1 itgens for chunk 30 since the chunks before 25 should be dropped. Got %v", itgens)
+	}
+}
+
+func itersToPoints(iters []tsz.Iter) []schema.Point {
+	points := make([]schema.Point, 0)
```
Comment (Dieterbe) [inline mdata/aggmetric_test.go]: https://dave.cheney.net/2013/01/19/what-is-the-zero-value-and-why-is-it-useful

Comment (fkaleo) [reply mdata/aggmetric_test.go]: Replaced with nil in commit ad7d980
